embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImpl.kt

```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * The span that models the lifetime of the current session or background activity
+     */
+    private val currentSessionSpan: AtomicReference<Span> = AtomicReference()
+
+    override fun startInitialSession(sdkInitStartTimeNanos: Long) {
+        synchronized(currentSessionSpan) {
+            currentSessionSpan.set(startSessionSpan(sdkInitStartTimeNanos))
+        }
+    }
+
+    /**
+     * Creating a new Span is only possible if the current session span is active, the parent has already been started, and the total
+     * session trace limit has not been reached. Once this method returns true, a new span is assumed to have been created and will
+     * be counted as such towards the limits, so make sure there's no case afterwards where a Span is not created.
+     */
+    override fun validateAndUpdateContext(parent: EmbraceSpan?, internal: Boolean): Boolean {
+        if (!currentSessionSpan.get().isRecording || (parent != null && parent.spanId == null)) {
+            return false
+        }
+
+        if (!internal) {
+            if (parent == null) {
+                if (currentSessionTraceCount.get() < SpansServiceImpl.MAX_TRACE_COUNT_PER_SESSION) {
+                    synchronized(currentSessionTraceCount) {
+                        if (currentSessionTraceCount.get() < SpansServiceImpl.MAX_TRACE_COUNT_PER_SESSION) {
+                            currentSessionTraceCount.incrementAndGet()
+                        } else {
+                            return false
+                        }
+                    }
+                } else {
+                    return false
+                }
+            } else {
+                val rootSpanId = getRootSpanId(parent)
+                val currentSpanCount = currentSessionChildSpansCount[rootSpanId]
+                if (currentSpanCount == null) {
+                    updateChildrenCount(rootSpanId)
+                } else if (currentSpanCount < SpansServiceImpl.MAX_SPAN_COUNT_PER_TRACE) {
+                    synchronized(currentSessionChildSpansCount) {
+                        val currentSpanCountAgain = currentSessionChildSpansCount[rootSpanId]
+                        if (currentSpanCountAgain == null || currentSpanCountAgain < SpansServiceImpl.MAX_SPAN_COUNT_PER_TRACE) {
+                            updateChildrenCount(rootSpanId)
+                        } else {
+                            return false
+                        }
+                    }
+                } else {
+                    return false
+                }
+            }
+        }
+
+        return true
+    }
+
+    override fun endSession(appTerminationCause: EmbraceAttributes.AppTerminationCause?): List<EmbraceSpanData> {
+        synchronized(this) {
+            // Right now, session spans don't survive native crashes and sudden process terminations,
+            // so telemetry will not be recorded in those cases, for now.
```

Comment:
Unrelated to the changes, but is there something on the roadmap to support this?

Reply:
Yes. We plan to cache active spans (which include session spans) periodically and have that cache be checked on start up next time when a crash session sent. Those spans will be terminated retroactively and the whole payload will be sent to the server as part of the session as span snapshot.

That's on the road map as part of the tracing GA work.